shared-libs/transitions/src/transitions/create_user_for_contacts.js

```diff
@@ -0,0 +1,127 @@
+const config = require('../config');
+const db = require('../db');
+const contactTypeUtils = require('@medic/contact-types-utils');
+const { people } = require('@medic/contacts')(config, db);
+const { users } = require('@medic/user-management')(config, db);
+
+const NAME = 'create_user_for_contacts';
+
+const UserCreationStatus = {
+  // PENDING - Set by webapp while waiting for sync to complete
+  READY: 'READY', // Ready to be replaced
+  COMPLETE: 'COMPLETE', // The new user has been created
+  ERROR: 'ERROR', // The new user could not be created
+};
+
+const getNewContact = (contactId) => {
+  if (!contactId) {
+    return Promise.reject(new Error('No id was provided for the new replacement contact.'));
+  }
+  return people.getOrCreatePerson(contactId);
+};
+
+const generateUsername = (contactName) => {
+  const randomNum = Math.floor(Math.random() * 9999).toString().padStart(4, '0');
+  const username = contactName.normalize('NFD') // split an accented letter in the base letter and the accent
+    .replace(/[\u0300-\u036f]/g, '') // remove all previously split accents
+    .toLowerCase()
+    .trim()
+    .replace(/[^a-z0-9 ]/g, '') // remove all chars not letters, numbers and spaces (to be replaced)
+    .replace(/\s+/g, '-'); // separator
+
+  return `${username}-${randomNum}`;
+};
+
+const generateUniqueUsername = (contactName) => {
+  const username = generateUsername(contactName);
+  return db.users.get(`org.couchdb.user:${username}`)
+    .then(() => generateUniqueUsername(contactName))
+    .catch(error => {
+      if (error.status === 404) {
+        // this username is available
+        return username;
+      }
+      throw error;
+    });
+};
+
+const createNewUser = ({ roles }, { _id, name, phone, parent }) => {
+  if (!name) {
+    throw new Error(`Replacement contact [${_id}] must have a name.`);
+  }
+  return generateUniqueUsername(name)
+    .then(username => {
+      const user = {
+        username,
+        token_login: true,
+        roles,
+        phone,
+        place: parent._id,
+        contact: _id,
+        fullname: name,
+      };
+      return users.createUser(user, config.get('app_url'))
+        .catch(err => {
+          if (!err.message || typeof err.message === 'string') {
+            throw err;
+          }
+          const message = err.message.message ? err.message.message : err.message;
+          throw new Error(`Error creating new user: ${JSON.stringify(message)}`);
+        });
+    });
+};
+
+const replaceUser = (originalContact) => {
+  const { _id, user_for_contact : { replaced: { by } } } = originalContact;
```

Comment:
I have updated this code to avoid just having the `by` variable on its own.  (Now it is `getNewContact(replaced.by)` instead of `getNewContact(by)`.

Let me know, though, if you actually meant that we should change the name of the field, itself, on the contact doc (so that it would be something like `user_for_contact.replaced.replaced_by`....)

Reply:
I generally think a two letter variable name that's also a preposition is probably less clear than using a full word or the prepositioned work along with the preposition. In this particular case, just using the "by" variable was weird. If the variable was `replacedBy`, this whole conversation would not have existed. 
If it were just up to me, I'd just rename the property to `replaced_by`, but I'm leaving that up to you.

Reply:
Changed to `user_for_contact.replace.replacement_contact_id`